src/components/DragAndDrop/index.js

```diff
@@ -123,72 +78,122 @@ class DragAndDrop extends React.Component {
+                            event.clientX >= dropZoneRect.current.right ||
+                            // Cancel drag when file manager is on top of the drop zone area - works only on chromium
+                            (event.target.getAttribute('id') === props.activeDropZoneId && !event.relatedTarget)
+                        ) {
+                            dropZoneDragState.current = DRAG_LEAVE_EVENT;
+                            props.onDragLeave(event);
+                        }
+                    }
+                    break;
+                case DROP_EVENT:
+                    dropZoneDragState.current = DRAG_LEAVE_EVENT;
+                    props.onDrop(event);
+                    break;
+                default:
+                    break;
+            }
+        },
+        [props, throttledDragOverHandler],
+    );
 
     /**
      * Handles all types of drag-N-drop events on the drop zone associated with composer
      *
      * @param {Object} event native Event
      */
-    dropZoneDragListener(event) {
-        event.preventDefault();
+    const dropZoneDragListener = useCallback(
+        (event) => {
+            event.preventDefault();
 
-        if (this.dropZone.contains(event.target) && this.props.shouldAcceptDrop(event)) {
-            this.dropZoneDragHandler(event);
+            if (dropZone.current.contains(event.target) && props.shouldAcceptDrop(event)) {
+                dropZoneDragHandler(event);
+            } else {
+                // eslint-disable-next-line no-param-reassign
+                event.dataTransfer.dropEffect = NONE_DROP_EFFECT;
+            }
+        },
+        [props, dropZoneDragHandler],
+    );
+
+    const addEventListeners = useCallback(() => {
+        dropZone.current = document.getElementById(props.dropZoneId);
+        dropZoneRect.current = calculateDropZoneClientReact();
+        document.addEventListener(DRAG_OVER_EVENT, dropZoneDragListener);
+        document.addEventListener(DRAG_ENTER_EVENT, dropZoneDragListener);
+        document.addEventListener(DRAG_LEAVE_EVENT, dropZoneDragListener);
+        document.addEventListener(DROP_EVENT, dropZoneDragListener);
+        window.addEventListener(RESIZE_EVENT, throttledDragNDropWindowResizeListener);
+    }, [props.dropZoneId, calculateDropZoneClientReact, dropZoneDragListener, throttledDragNDropWindowResizeListener]);
+
+    const removeEventListeners = useCallback(() => {
+        document.removeEventListener(DRAG_OVER_EVENT, dropZoneDragListener);
+        document.removeEventListener(DRAG_ENTER_EVENT, dropZoneDragListener);
+        document.removeEventListener(DRAG_LEAVE_EVENT, dropZoneDragListener);
+        document.removeEventListener(DROP_EVENT, dropZoneDragListener);
+        window.removeEventListener(RESIZE_EVENT, throttledDragNDropWindowResizeListener);
+    }, [dropZoneDragListener, throttledDragNDropWindowResizeListener]);
+
+    useEffect(() => {
```

Comment:
Good point! I also thought that this was a bit weird, but decided to keep the current implementation since the purpose of the task is just to rewrite the component from class to functional syntax. I'll implement the changes you suggested.